cpp/docs/DEVELOPER_GUIDE.md

```diff
@@ -0,0 +1,973 @@
+when a non-pointer value is returned from the API that is the result of an asynchronous 
+device-to-host copy, the stream used for the copy should be synchronized before returning. However,
+when a column is returned, the stream should not be synchronized because doing so will break 
+asynchrony if and when we add an asynchronous API to libcudf.
+
+**Note:** `cudaDeviceSynchronize()` should *never* be used.
+ This limits the ability to do any multi-stream/multi-threaded work with libcudf APIs.
+
+ ### Stream Creation
+
+There may be times in implementing libcudf features where it would be advantageous to use streams 
+*internally*, i.e., to accomplish overlap in implementing an algorithm. However, dynamically 
+creating a stream can be expensive. RMM has recently added a stream pool for this situation to 
+avoid dynamic stream creation. However, this is not yet exposed in libcudf, so for the time being, 
+libcudf features should avoid creating streams (even if it is slightly less efficient). It is a 
+good idea to leave a `// TODO:` note indicating where using a stream would be beneficial.
+
+## Memory Allocation
+
+Device [memory resources](#memory_resource) are used in libcudf to abstract and control how device 
+memory is allocated. 
+
+### Output Memory
+
+Any libcudf API that allocates memory that is *returned* to a user must accept a pointer to a 
+`device_memory_resource` as the last parameter. Inside the API, this memory resource must be used
+to allocate any memory for returned objects. It should therefore be passed into functions whose
+outputs will be returned. Example:
+
+```c++
+// Returned `column` contains newly allocated memory, 
+// therefore the API must accept a memory resource pointer
+std::unique_ptr<column> returns_output_memory(
+  ..., rmm::device_memory_resource * mr = rmm::mr::get_current_device_resource());
+
+// This API does not allocate any new *output* memory, therefore
+// a memory resource is unnecessary
+void does_not_allocate_output_memory(...);                                              
+```
+
+### Temporary Memory
+
+Not all memory allocated within a libcudf API is returned to the caller. Often algorithms must 
+allocate temporary, scratch memory for intermediate results. Always use the default resource
+obtained from `rmm::mr::get_current_device_resource()` for temporary memory allocations. Example:
+
+```c++
+rmm::device_buffer some_function(
+  ..., rmm::mr::device_memory_resource mr * = rmm::mr::get_current_device_resource()) {
+    rmm::device_buffer returned_buffer(..., mr); // Returned buffer uses the passed in MR
+    ...
+    rmm::device_buffer temporary_buffer(...); // Temporary buffer uses default MR
+    ...
+    return returned_buffer;
+}
+```
+
+### Memory Management
+
+libcudf code generally eschews raw pointers and direct memory allocation. Use RMM classes built to
+use `device_memory_resource`(*)s for device memory allocation with automated lifetime management.
+
+#### `rmm::device_buffer`
```

Comment:
Maybe these classes should be defined in the rmm repo developer guide and then just pointed to from here?

Reply:
We don't have such a guide yet. But when we do...